### PR TITLE
Fixed test session issues

### PR DESCRIPTION
**Pull request by cvpoienaru**
## Description
This PR fixes a couple of issues related to test sessions:
- Spawning a test session when either the runsettings or the test sources are incompatible with test session criteria (e.g.: the use of Fakes or any other data collector);
- Aborting test discovery/run operations because of an unhandled exception when returning a non-existent proxy, caused by an earlier incompatibility between the session criteria when the session was spawned and the current operation criteria. This scenario was taken into account before and the solution was to spawn an on-demand testhost that meets the discovery/run criteria, but it wasn't properly tested;
- During the abort scenario described above, a testhost Close call is badly handled and has the effect of leaving on-demand testhosts hanging indefinitely. This issue has been addressed by the current PR, too;

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs [hunk 1]
```diff
@@ -238,13 +238,17 @@ public void Close()
...
         //
         // In contrast, the new workflow (using test sessions) means we should keep
-        // the testhost alive until explicitly closed by the test session owner.
-        if (_testSessionInfo == null)
+        // the testhost alive until explicitly closed by the test session owner, but
+        // only if the testhost is part of a test session (i.e. the proxy operation manager
+        // id is valid), since there is the distinct possibility of test session criteria
+        // changing between spawn and discovery/run, causing a new proxy operation manager
+        // to be spawned on demand instead of dequing an incompatible proxy from the pool.
+        if (_testSessionInfo == null || _proxyOperationManager.Id < 0)
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs:251]: Not sure to fully understand (based on comment). Are you saying that we except the ID of the proxy operation manager to be always negative when we are ready to close?

**Review by Evangelink — COMMENTED** [review]
A few comments and nits. I am not asking for tests as you told me that you looked into it and it's not easy to add them because of current architecture.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs [hunk 2]
```diff
@@ -238,13 +238,17 @@ public void Close()
...
         // successfully completing the operation it was spawned for.
         //
         // In contrast, the new workflow (using test sessions) means we should keep
-        // the testhost alive until explicitly closed by the test session owner.
-        if (_testSessionInfo == null)
+        // the testhost alive until explicitly closed by the test session owner, but
+        // only if the testhost is part of a test session (i.e. the proxy operation manager
+        // id is valid), since there is the distinct possibility of test session criteria
+        // changing between spawn and discovery/run, causing a new proxy operation manager
+        // to be spawned on demand instead of dequing an incompatible proxy from the pool.
```
[the remark this answers is not in the document]
Comment (cvpoienaru) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs:250]: Fixed, thanks !

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 3]
```diff
@@ -854,6 +847,39 @@ static Architecture TranslateToArchitecture(PlatformArchitecture targetArchitect
...
+        // because why would they be? We never pass that forward to the executor, so this probably should
+        // just look at sources anyway.
+        //
+        // The commandline options do not have sources in design time mode,
+        // and so we fall back to using sources instead.
+        if (_commandLineOptions.Sources.Any())
+        {
+            GenerateFakesUtilities.GenerateFakesSettings(
+                _commandLineOptions,
+                _commandLineOptions.Sources.ToList(),
```
[the remark this answers is not in the document]
Comment (cvpoienaru) [reply src/vstest.console/TestPlatformHelpers/TestRequestManager.cs:869]: Done.

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 4]
```diff
@@ -854,6 +847,39 @@ static Architecture TranslateToArchitecture(PlatformArchitecture targetArchitect
...
+                ref runsettings);
+        }
+        else if (sources.Any())
```
[the remark this answers is not in the document]
Comment (cvpoienaru) [reply src/vstest.console/TestPlatformHelpers/TestRequestManager.cs:872]: Done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/TestSessionPool.cs [hunk 5]
```diff
@@ -198,7 +198,7 @@ public virtual bool ReturnProxy(TestSessionInfo testSessionInfo, int proxyId)
             // Try re-enqueueing the specified proxy.
             return sessionManager.EnqueueProxy(proxyId);
         }
-        catch (InvalidOperationException ex)
+        catch (Exception ex)
```
[the remark this answers is not in the document]
Comment (cvpoienaru) [reply src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/TestSessionPool.cs:201]: We do indeed need to catch generic exceptions here. I'll add a comment in the code with the same reasoning I'm sharing right now:

Returning a proxy is more like a fire-and-forget kind of operation, it's no biggie if we aren't able to return it for whatever reason. In fact, this happens on a regular basis since there's 2 distinct calls to `ProxyDiscoveryManager/ProxyExecutionManager.Close()`, one when handling the raw message, and one when processing an actual termination message. Only one of those operations will succeed, the other one will fail, and there's no way around it without an architectural change. That's why we used to catch only `InvalidOperationException`, to account for the edge case outlined above. However, things changed when we added the fallback capability for test sessions (i.e. if we cannot match the discovery/run criteria with the criteria from one active test session that already has pre-started testhosts, then we'll just spawn testhosts on demand). The on-demand testhosts have the same regime as older testhosts, i.e. they just die after the operation they were spawned for has completed. The difference between those testhosts and the ones that are managed by a test session is that they have "-1" for the Id, and if by mistake we attempt to return one of those "non-managed" testhosts to a test session, we throw an exception that will not be caught by this method if we only handle `InvalidOperationException`. As such, rather then rely on the caller to perform sanity checks, we make sure such an edge case will not throw instead.

[the remark this answers is not in the document]
Comment (cvpoienaru) [reply src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/TestSessionPool.cs:201]: Fixed, added the comment.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs (hunk 1 above)
Comment (cvpoienaru) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs:251]: Yep. A negative ID means the proxy operation manager is not part of a test session, and hence can be disposed of here (both no-session discovery/run and on-demand proxy spawn fit in this category). If the ID is non-negative, this means the proxy is managed by a session and it will be closed only when explicitly requested by the session creator.

**Review by Evangelink — APPROVED** [review]
LGTM, might be worth asking for a second opinion.

**Comment by cvpoienaru** [comment]
To be honest, I'd rather just merge since the PR has been up for review since Monday and nobody else weighed in.
